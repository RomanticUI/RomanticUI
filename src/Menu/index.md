---
nav: 组件
group: 导航
mobile: false
toc: content
---

# Menu

## 代码演示

```tsx
import { AllItems, Menu } from 'RomanticUI';
const item: AllItems = [
  {
    label: 'Navigation One',
    optionKey: 'mail',
  },
  {
    label: 'Navigation Two',
    optionKey: 'app',
  },
  {
    label: 'Navigation Three - Submenu-kkkkkkk',
    optionKey: 'SubMenu',
    children: [
      {
        type: 'group',
        label: 'Item 1',
        children: [
          {
            label: 'Option 1',
            optionKey: 'setting:1',
          },
          {
            label: 'Option 2',
            optionKey: 'setting:2',
          },
        ],
      },
      {
        type: 'group',
        label: 'Item 2',
        children: [
          {
            label: 'Option 3',
            optionKey: 'setting:3',
          },
          {
            label: 'Option 4',
            optionKey: 'setting:4',
          },
        ],
      },
    ],
  },
  {
    label: 'alipay',
    optionKey: 'alipay',
  },
];
export default () => (
  <div>
    <Menu items={item}></Menu>
  </div>
);
```

## API

| 属性  | 说明           | 类型     | 默认值 |
| ----- | -------------- | -------- | ------ |
| items | 设置 Menu 内容 | AllItems | null   |
