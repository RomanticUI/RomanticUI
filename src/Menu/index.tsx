import React, { createContext, useEffect, useRef, useState } from 'react';
import './style/index.less';

interface SingleItem {
  label: string;
  optionKey?: string;
  icon?: any;
  children?: AllItems;
  type?: string;
  disable?: boolean;
}

type AllItems = Array<SingleItem>;

type MenuProps = {
  mode?: 'vertical' | 'horizontal';
  items: AllItems;
  isTop?: boolean;
  theme?: 'light' | 'night';
};

const SubMenu: React.FC<SingleItem> = ({
  label,
  optionKey,
  children,
  type,
  disable = false,
  ...others
}: SingleItem) => {
  const [sonCheck, setSoncheck] = useState<boolean>(false);

  /* 使用ref绑定span */
  const spanRef = useRef<HTMLSpanElement>(null);

  return (
    <Consumer>
      {(value) => {
        return (
          /* 判断是否是禁用样式 是否为group */
          <li
            key={optionKey}
            className={
              disable
                ? type == 'group'
                  ? 'disable group'
                  : 'disable'
                : type == 'group'
                  ? ' group'
                  : ''
            }
          >
            {/* 如果是disable 生成无交互元素 */}

            {disable ? (
              <div className="main">
                <span>{label}</span>
              </div>
            ) : (
              <div
                className="main"
                onClick={() => {
                  value.changeKey_Sub!(optionKey!);
                }}
              >
                <span
                  ref={spanRef}
                  className={optionKey == value.keyCked_Sub ? 'subSpanClicked' : ''}
                >
                  {label}
                </span>
                <ul className={'subUl' + ' ' + value.theme}>
                  {children
                    ? children!.map((curr, index) => {
                      //   console.log(curr.key);
                      return curr.children ? (
                        // 如果type为group 就不生成span,并且li类为group
                        curr.type == 'group' ? (
                          <SubMenu {...curr} key={curr.label} />
                        ) : (
                          /* 否则生成span */
                          <SubMenu {...curr} key={curr.label} />
                        )
                      ) : (
                        /* 传入handle函数，当子组件被点击后通知父组件 */
                        <MenuItem {...curr} key={curr.label} />
                      );
                    })
                    : ''}
                </ul>
              </div>
            )}
          </li>
        );
      }}
    </Consumer>
  );
};

const MenuItem: React.FC<SingleItem> = (props) => {
  let { label, optionKey, disable = false, ...others } = props;

  /* 鼠标点击后的处理函数 */
  type ConsumerValue = {
    keyCked_Item: string;
    changeKey_Item: React.Dispatch<React.SetStateAction<string>> | null;
  };
  const ItemClicked: ({ changeKey_Item }: ConsumerValue) => void = ({ changeKey_Item }) => {
    changeKey_Item!(optionKey!);
  };

  // console.log(props);
  return (
    <Consumer>
      {(value) => {
        return (
          /* 如果是disable 生成无交互元素 */
          disable ? (
            <li className="disable">
              <span>{label}</span>
            </li>
          ) : (
            <li
              key={optionKey!}
              className={optionKey == value.keyCked_Item ? 'ItemClicked ItemLi' : 'ItemLi'}
              onClick={() => ItemClicked(value)}
            >
              <span>{label}</span>
            </li>
          )
        );
      }}
    </Consumer>
  );
};

const { Provider, Consumer } = createContext<{
  keyCked_Item: string;
  changeKey_Item: React.Dispatch<React.SetStateAction<string>> | null;
  keyCked_Sub: string;
  changeKey_Sub: React.Dispatch<React.SetStateAction<string>> | null;
  themeStyle?: {
    backcolor: string;
    clickedFontClor: string;
    clickedBackClor: string;
  };
  theme: 'light' | 'night';
}>({} as any);

const Menu: React.FC<MenuProps> = ({
  items,
  isTop = true,
  theme = 'light',
  mode = 'horizontal',
  ...others
}: MenuProps) => {
  useEffect(() => { }, []);

  const [keyCked_Item, changeKey_Item] = useState('');
  const [keyCked_Sub, changeKey_Sub] = useState('');

  return (
    <Provider value={{ keyCked_Item, changeKey_Item, keyCked_Sub, changeKey_Sub, theme }}>
      <div className="MenuBox main">
        {/* Menu模式 以及Menu主题 */}
        <ul className={'subUl' + ' ' + theme + ' ' + mode + ' ' + 'top'}>
          {items.map((curr, index) => {
            // console.log({ ...curr });
            return <SubMenu {...curr} key={index} />;
          })}
        </ul>
      </div>
    </Provider>
  );
};

export default Menu;
export type { AllItems };
