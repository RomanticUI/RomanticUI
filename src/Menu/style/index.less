//menu默认导航样式：顶级导航
.MenuBox {
  box-sizing: border-box;
  ul {
    width: 100%;
    list-style: none; //取消默认装饰
  }
  li {
    height: 45px;
    list-style: none; //取消默认装饰
    cursor: pointer; //鼠标样式指针
  }
  span {
    display: inline-block;
    width: auto;
  }
  div {
    width: auto;
  }
}

.main {
  //组件基本样式
  li {
    width: auto;
    // flex-grow: 1;
    height: auto;
    padding: 0 8px;
    line-height: 45px;
    > div {
      width: 100%;
      height: 100%;
    }
  }
}

.MenuBox {
  width: auto;
}

/* SubMenu并且horizontal */
.subUl.horizontal {
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  width: auto;
  margin: 0;
  padding: 0;
  .subUl {
    width: 100%;
  }

  //topUl下的首个Menu要绝对定位
  > li {
    width: auto;
    > div {
      //子绝父相
      position: relative;

      > ul {
        position: absolute;
        left: 0;
        z-index: 1;
        //隐藏
        transform: scaleY(-0);
        opacity: 0;
        transition: ease-in-out 0.1s;
      }

      > div {
        height: auto;
        color: bisque;
      }
      &:hover {
        > ul {
          transform: scaleY(1);
          opacity: 1;
        }
      }
    }
  }
}

/* SubMenu并且top并且vertical */
.subUl {
  width: 200px;
}

/* SubMenu并且vertical*/
.subUl.vertical.top {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  justify-content: flex-start;
  width: 300px;

  margin: 0;
  padding: 0;

  span {
    width: 100%;
  }

  //topUl下的首个Menu要绝对定位
  > li {
    width: 100%;
    > div {
      //子绝父相
      position: relative;
      > span {
        position: relative;
        width: 100%;
        &::after {
          position: absolute;
          top: 0;
          right: 10px;
          content: '>';
        }
      }
      > ul {
        position: absolute;
        top: 0px;
        left: 105%;
        z-index: 1;
        //隐藏
        transform: scale(-0);
        opacity: 0;
        transition: ease-in-out 0.2s;
      }

      > div {
        height: auto;
        color: bisque;
      }
      &:hover {
        > ul {
          transform: scale(1, 1);
          opacity: 1;
        }
      }
    }
  }
}

/* 项目 悬浮变色 */
.subUl {
  border-radius: 3px;
  > li:hover {
    background-color: rgba(165, 165, 165, 0.3);
  }
}

//二级menu样式：submenu
.subUl {
  padding: 0;
}

/* group类型 */
.group {
  > div {
    width: 100% !important;
    > span {
      color: rgba(138, 138, 138, 0.8);
    }
  }

  /* 让group下的subUl适应父元素大小 */
  .subUl {
    width: 100% !important;
  }
}

.subSpanClicked {
  color: rgb(57, 121, 207);
}

/* theme相关 */
.night {
  color: rgba(255, 255, 255, 0.65);
  background-color: rgb(4, 16, 37);
  .subSpanClicked {
    color: white;
    &::after {
      color: white;
    }
  }
  .subUl {
    padding: 0;
  }

  /* group li悬浮无效果 */
  .group:hover {
    background-color: rgb(4, 16, 37);
  }

  .ItemLi {
    background-color: rgb(4, 16, 37);
  }

  .ItemClicked {
    color: white;
    background-color: rgb(57, 121, 207);
    &:hover {
      background-color: rgb(57, 121, 207);
    }
  }

  /* disable样式 */
  .disable {
    color: gray;
    cursor: not-allowed;
    &:hover {
      background-color: rgb(4, 16, 37);
    }
  }
}

.ItemLi {
  border-radius: 5px;
}
.subUl {
  border-radius: 5px;
}
.light {
  color: black;
  background-color: rgb(255, 255, 255);

  .subSpanClicked {
    color: rgb(57, 121, 207);
    &::after {
      color: rgb(57, 121, 207);
    }
  }

  /* group li悬浮无效果 */
  .group:hover {
    background-color: rgb(255, 255, 255);
  }

  .ItemClicked {
    color: rgb(57, 121, 207);
    background-color: rgba(102, 156, 220, 0.4);
    &:hover {
      background-color: rgba(102, 156, 220, 0.4);
    }
  }
  /* disable样式 */
  .disable {
    color: rgba(0, 0, 0, 0.2);
    cursor: not-allowed;
    &:hover {
      background-color: white;
    }
  }
}
