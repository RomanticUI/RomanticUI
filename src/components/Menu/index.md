---
nav: 组件
group: 导航
mobile: false
toc: content
---

# Menu

## 代码演示

### 基本

```tsx
import { AllItems, Menu } from 'RomanticUI';
const item: AllItems = [
  {
    label: 'Navigation One',
    optionKey: 'mail',
  },
  {
    label: 'Navigation Two',
    optionKey: 'app',
    children: [
      {
        label: 'Option 1',
        optionKey: 'setting:1',
      },
      {
        label: 'Option 2',
        optionKey: 'setting:2',
      },
      {
        label: 'Option 3',
        optionKey: 'setting:3',
      },
    ],
  },
  {
    label: 'Navigation Three - Submenu-kkkkkkk',
    optionKey: 'SubMenu',
    children: [
      {
        type: 'group',
        label: 'Item 1',
        children: [
          {
            label: 'Option 4',
            optionKey: 'setting:4',
          },
          {
            label: 'Option 5',
            optionKey: 'setting:5',
          },
        ],
      },
      {
        type: 'group',
        label: 'Item 2',
        children: [
          {
            label: 'Option 6',
            optionKey: 'setting:6',
          },
          {
            label: 'Option 7',
            optionKey: 'setting:7',
          },
        ],
      },
    ],
  },
  {
    label: 'alipay',
    optionKey: 'alipay',
  },
];
export default () => (
  <div>
    <Menu items={item}></Menu>
  </div>
);
```

### 主题

有 light 和 night 主题，默认 light

```tsx
import { AllItems, Menu } from 'RomanticUI';
const item: AllItems = [
  {
    label: 'Navigation One',
    optionKey: 'mail',
  },
  {
    label: 'Navigation Two',
    optionKey: 'app',
    children: [
      {
        label: 'Option 1',
        optionKey: 'setting:1',
      },
      {
        label: 'Option 2',
        optionKey: 'setting:2',
      },
      {
        label: 'Option 3',
        optionKey: 'setting:3',
      },
    ],
  },
  {
    label: 'Navigation Three - Submenu-kkkkkkk',
    optionKey: 'SubMenu',
    children: [
      {
        type: 'group',
        label: 'Item 1',
        children: [
          {
            label: 'Option 4',
            optionKey: 'setting:4',
          },
          {
            label: 'Option 5',
            optionKey: 'setting:5',
          },
        ],
      },
      {
        type: 'group',
        label: 'Item 2',
        children: [
          {
            label: 'Option 6',
            optionKey: 'setting:6',
          },
          {
            label: 'Option 7',
            optionKey: 'setting:7',
          },
        ],
      },
    ],
  },
  {
    label: 'alipay',
    optionKey: 'alipay',
  },
];
export default () => (
  <div>
    <Menu items={item} theme="night"></Menu>
  </div>
);
```

### 禁用某项

在属性 Items 中设置 `disable:true` 可将该项导航禁用（默认为 true 不禁用）

```tsx
import { AllItems, Menu } from 'RomanticUI';
const item: AllItems = [
  {
    label: 'Navigation One',
    optionKey: 'mail',
  },
  {
    label: 'Navigation Two',
    optionKey: 'app',
    disable: true,
    children: [
      {
        label: 'Option 1',
        optionKey: 'setting:1',
      },
      {
        label: 'Option 2',
        optionKey: 'setting:2',
      },
      {
        label: 'Option 3',
        optionKey: 'setting:3',
      },
    ],
  },
  {
    label: 'Navigation Three - Submenu-kkkkkkk',
    optionKey: 'SubMenu',
    children: [
      {
        type: 'group',
        label: 'Item 1',
        children: [
          {
            label: 'Option 4',
            optionKey: 'setting:4',
          },
          {
            label: 'Option 5',
            optionKey: 'setting:5',
            disable: true,
          },
        ],
      },
      {
        type: 'group',
        label: 'Item 2',
        children: [
          {
            label: 'Option 6',
            optionKey: 'setting:6',
          },
          {
            label: 'Option 7',
            optionKey: 'setting:7',
          },
        ],
      },
    ],
  },
  {
    label: 'alipay',
    optionKey: 'alipay',
  },
];
export default () => (
  <div>
    <Menu items={item} theme="night"></Menu>
  </div>
);
```

### Menu 模式

可设置水平或垂直菜单

```tsx
import { AllItems, Menu } from 'RomanticUI';
const item: AllItems = [
  {
    label: 'Navigation One',
    optionKey: 'mail',
  },
  {
    label: 'Navigation Two',
    optionKey: 'app',
    children: [
      {
        label: 'Option 1',
        optionKey: 'setting:1',
      },
      {
        label: 'Option 2',
        optionKey: 'setting:2',
      },
      {
        label: 'Option 3',
        optionKey: 'setting:3',
      },
    ],
  },
  {
    label: 'Navigation Three - Submenu',
    optionKey: 'SubMenu',
    children: [
      {
        type: 'group',
        label: 'Item 1',
        children: [
          {
            label: 'Option 4',
            optionKey: 'setting:4',
          },
          {
            label: 'Option 5',
            optionKey: 'setting:5',
          },
        ],
      },
      {
        type: 'group',
        label: 'Item 2',
        children: [
          {
            label: 'Option 6',
            optionKey: 'setting:6',
          },
          {
            label: 'Option 7',
            optionKey: 'setting:7',
          },
        ],
      },
    ],
  },
  {
    label: 'alipay',
    optionKey: 'alipay',
  },
];
export default () => (
  <div>
    <Menu items={item} mode="vertical" theme="night"></Menu>
  </div>
);
```

## API

### Menu

| 属性  | 说明                       | 类型                       | 默认值     |
| ----- | -------------------------- | -------------------------- | ---------- |
| items | 设置 Menu 内容             | AllItems                   | null       |
| theme | 设置 Menu 主题             | "light" \| "night"         | "light"    |
| mode  | 设置 Menu 模式，水平或垂直 | "vertical" \| "horizontal" | "vertical" |

### items

| 属性      | 说明                             | 类型     | 默认值 |
| --------- | -------------------------------- | -------- | ------ |
| label     | 该项文本                         | String   | null   |
| optionKey | 该项的 id，用于交互              | string   | null   |
| type      | 该项类型：group 类型无 optionKey | string   | ""     |
| disable   | 是否禁用该项                     | boolean  | false  |
| children  | 该项的子菜单                     | AllItems | null   |
