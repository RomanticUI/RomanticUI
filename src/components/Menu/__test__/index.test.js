// 引入测试组件
import React from 'react';
import Menu from '..';
// 使用@testing-library/react渲染对应组件
import { render } from '@testing-library/react';

describe('Divider test', () => {
  it('test case1 for night', () => {
    const item = [
      {
        label: 'Navigation One',
        optionKey: 'mail',
      },
      {
        label: 'Navigation Two',
        optionKey: 'app',
        children: [
          {
            label: 'Option 1',
            optionKey: 'setting:1',
          },
          {
            label: 'Option 2',
            optionKey: 'setting:2',
          },
          {
            label: 'Option 3',
            optionKey: 'setting:3',
          },
        ],
      },
      {
        label: 'Navigation Three - Submenu-kkkkkkk',
        optionKey: 'SubMenu',
        children: [
          {
            type: 'group',
            label: 'Item 1',
            children: [
              {
                label: 'Option 4',
                optionKey: 'setting:4',
              },
              {
                label: 'Option 5',
                optionKey: 'setting:5',
              },
            ],
          },
          {
            type: 'group',
            label: 'Item 2',
            children: [
              {
                label: 'Option 6',
                optionKey: 'setting:6',
              },
              {
                label: 'Option 7',
                optionKey: 'setting:7',
              },
            ],
          },
        ],
      },
      {
        label: 'alipay',
        optionKey: 'alipay',
      },
    ];
    const { container } = render(<Menu items={item} theme="night"></Menu>);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('test case1 for solid', () => {
    const item = [
      {
        label: 'Navigation One',
        optionKey: 'mail',
      },
      {
        label: 'Navigation Two',
        optionKey: 'app',
        children: [
          {
            label: 'Option 1',
            optionKey: 'setting:1',
          },
          {
            label: 'Option 2',
            optionKey: 'setting:2',
          },
          {
            label: 'Option 3',
            optionKey: 'setting:3',
          },
        ],
      },
      {
        label: 'Navigation Three - Submenu',
        optionKey: 'SubMenu',
        children: [
          {
            type: 'group',
            label: 'Item 1',
            children: [
              {
                label: 'Option 4',
                optionKey: 'setting:4',
              },
              {
                label: 'Option 5',
                optionKey: 'setting:5',
              },
            ],
          },
          {
            type: 'group',
            label: 'Item 2',
            children: [
              {
                label: 'Option 6',
                optionKey: 'setting:6',
              },
              {
                label: 'Option 7',
                optionKey: 'setting:7',
              },
            ],
          },
        ],
      },
      {
        label: 'alipay',
        optionKey: 'alipay',
      },
    ];
    const { container } = render(<Menu items={item} mode="vertical" theme="night"></Menu>);

    expect(container.firstChild).toMatchSnapshot();
  });
});
