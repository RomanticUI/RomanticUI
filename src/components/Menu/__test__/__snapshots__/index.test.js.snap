// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Divider test test case1 for night 1`] = `
<div
  class="MenuBox main"
>
  <ul
    class="subUl night horizontal top"
  >
    <li
      class=""
    >
      <div
        class="main"
      >
        <span
          class=""
        >
          Navigation One
        </span>
        <ul
          class="subUl night"
        />
      </div>
    </li>
    <li
      class=""
    >
      <div
        class="main"
      >
        <span
          class=""
        >
          Navigation Two
        </span>
        <ul
          class="subUl night"
        >
          <li
            class="ItemLi"
          >
            <span>
              Option 1
            </span>
          </li>
          <li
            class="ItemLi"
          >
            <span>
              Option 2
            </span>
          </li>
          <li
            class="ItemLi"
          >
            <span>
              Option 3
            </span>
          </li>
        </ul>
      </div>
    </li>
    <li
      class=""
    >
      <div
        class="main"
      >
        <span
          class=""
        >
          Navigation Three - Submenu-kkkkkkk
        </span>
        <ul
          class="subUl night"
        >
          <li
            class=" group"
          >
            <div
              class="main"
            >
              <span
                class=""
              >
                Item 1
              </span>
              <ul
                class="subUl night"
              >
                <li
                  class="ItemLi"
                >
                  <span>
                    Option 4
                  </span>
                </li>
                <li
                  class="ItemLi"
                >
                  <span>
                    Option 5
                  </span>
                </li>
              </ul>
            </div>
          </li>
          <li
            class=" group"
          >
            <div
              class="main"
            >
              <span
                class=""
              >
                Item 2
              </span>
              <ul
                class="subUl night"
              >
                <li
                  class="ItemLi"
                >
                  <span>
                    Option 6
                  </span>
                </li>
                <li
                  class="ItemLi"
                >
                  <span>
                    Option 7
                  </span>
                </li>
              </ul>
            </div>
          </li>
        </ul>
      </div>
    </li>
    <li
      class=""
    >
      <div
        class="main"
      >
        <span
          class=""
        >
          alipay
        </span>
        <ul
          class="subUl night"
        />
      </div>
    </li>
  </ul>
</div>
`;

exports[`Divider test test case1 for solid 1`] = `
<div
  class="MenuBox main"
>
  <ul
    class="subUl night vertical top"
  >
    <li
      class=""
    >
      <div
        class="main"
      >
        <span
          class=""
        >
          Navigation One
        </span>
        <ul
          class="subUl night"
        />
      </div>
    </li>
    <li
      class=""
    >
      <div
        class="main"
      >
        <span
          class=""
        >
          Navigation Two
        </span>
        <ul
          class="subUl night"
        >
          <li
            class="ItemLi"
          >
            <span>
              Option 1
            </span>
          </li>
          <li
            class="ItemLi"
          >
            <span>
              Option 2
            </span>
          </li>
          <li
            class="ItemLi"
          >
            <span>
              Option 3
            </span>
          </li>
        </ul>
      </div>
    </li>
    <li
      class=""
    >
      <div
        class="main"
      >
        <span
          class=""
        >
          Navigation Three - Submenu
        </span>
        <ul
          class="subUl night"
        >
          <li
            class=" group"
          >
            <div
              class="main"
            >
              <span
                class=""
              >
                Item 1
              </span>
              <ul
                class="subUl night"
              >
                <li
                  class="ItemLi"
                >
                  <span>
                    Option 4
                  </span>
                </li>
                <li
                  class="ItemLi"
                >
                  <span>
                    Option 5
                  </span>
                </li>
              </ul>
            </div>
          </li>
          <li
            class=" group"
          >
            <div
              class="main"
            >
              <span
                class=""
              >
                Item 2
              </span>
              <ul
                class="subUl night"
              >
                <li
                  class="ItemLi"
                >
                  <span>
                    Option 6
                  </span>
                </li>
                <li
                  class="ItemLi"
                >
                  <span>
                    Option 7
                  </span>
                </li>
              </ul>
            </div>
          </li>
        </ul>
      </div>
    </li>
    <li
      class=""
    >
      <div
        class="main"
      >
        <span
          class=""
        >
          alipay
        </span>
        <ul
          class="subUl night"
        />
      </div>
    </li>
  </ul>
</div>
`;
